riscv_executor/src/lib.rs

```diff
@@ -0,0 +1,714 @@
+                | FunctionStatement::Return(_) => {
+                    statement_seen = true;
+                    statements.push(s)
+                }
+                FunctionStatement::DebugDirective(d) => {
+                    match &d.directive {
+                        DebugDirective::File(idx, dir, file) => {
+                            // debug files should be densely packed starting
+                            // from 1, so the idx should match vec size + 1:
+                            assert_eq!(*idx, debug_files.len() + 1);
+                            debug_files.push((dir.as_str(), file.as_str()));
+                        }
+                        DebugDirective::Loc(_, _, _) | DebugDirective::OriginalInstruction(_) => {
+                            // keep debug locs for debugging purposes
+                            statements.push(s);
+                        }
+                    }
+                }
+                FunctionStatement::Label(LabelStatement { start: _, name }) => {
+                    // assert there are no statements in the middle of a block
+                    assert!(!statement_seen);
+                    label_map.insert(name.as_str(), (batch_idx as i64).into());
+                }
+            }
+        }
+    }
+    assert!(statements.len() <= u32::MAX as usize);
+
+    // add a final element to the map so the queries don't overflow:
+    batch_to_line_map.push(statements.len() as u32);
+
+    PreprocessedMain {
+        statements,
+        label_map,
+        batch_to_line_map,
+        debug_files,
+    }
+}
+
+struct Executor<'a, 'b, F: FieldElement> {
+    proc: TraceBuilder<'a, 'b>,
+    mem: MemoryBuilder,
+    label_map: HashMap<&'a str, Elem>,
+    inputs: &'b [F],
+    stdout: io::Stdout,
+}
+
+impl<'a, 'b, F: FieldElement> Executor<'a, 'b, F> {
+    fn exec_instruction(&mut self, name: &str, args: &[Expression<F>]) -> Vec<Elem> {
+        let args = args
+            .iter()
+            .map(|expr| self.eval_expression(expr)[0])
+            .collect::<Vec<_>>();
+
+        match name {
+            "mstore" => {
+                let addr = args[0].0 as u32;
+                assert_eq!(addr % 4, 0);
+                self.mem.s(args[0].0 as u32, args[1]);
+
+                Vec::new()
+            }
+            "mload" => {
+                let addr = args[0].0 as u32;
+                let val = self.mem.g(addr & 0xfffffffc);
+                let rem = addr % 4;
+
+                vec![val, rem.into()]
+            }
+            "jump" => {
+                self.proc.s("pc", args[0]);
+
+                Vec::new()
+            }
+            "load_label" => args,
+            "jump_dyn" => {
+                self.proc.s("pc", args[0]);
+
+                Vec::new()
+            }
+            "jump_and_link_dyn" => {
+                let pc = self.proc.g("pc");
+                self.proc.s("x1", pc.u() + 1);
+                self.proc.s("pc", args[0]);
```

Comment:
Maybe also add a `set_pc` function? Is it actually legal to set the PC in a regular instruction? If not, we could remove the special case in the `s` function.

Reply:
You mean, like in `addi pc,pc,42`? Some random interwebs slides says it is not possible. I'll do it.

But still find the `s(key, value)` solution more elegant than returning `mut &`.

Reply:
Done.